ts/packages/node/src/stores/LaunchDarkly.ts

```diff
@@ -0,0 +1,173 @@
+      },
+      key: config.key,
+      name: config.key,
+      variations: [{ value: config.value }, { value: '' }],
+    };
+    const { data: featureFlag } = await this.client.post(`/flags/${this.projectKey}`, createData);
+    return featureFlag;
+  }
+
+  private async createVariation(config: Config) {
+    const createVariationData = {
+      instructions: [{ kind: 'addVariation', value: config.value }],
+    };
+    const data = await this.patchUpdate(config, createVariationData);
+    return data.variations.find((variation: any) => variation.value === config.value);
+  }
+
+  private async updateDefaultFallthroughVariation(config: Config, variationId: string) {
+    const updateDefaultVariationData = {
+      instructions: [
+        {
+          kind: 'updateDefaultVariation',
+          OnVariationValue: config.value,
+          OffVariationValue: '',
+        },
+      ],
+    };
+    const updateFallthroughVariationData = {
+      environmentKey: config.set,
+      instructions: [
+        {
+          kind: 'updateFallthroughVariationOrRollout',
+          variationId,
+        },
+      ],
+    };
+    await this.patchUpdate(config, updateDefaultVariationData);
+    return this.patchUpdate(config, updateFallthroughVariationData);
+  }
+
+  private async upsert(config: Config) {
+    if (!this.environments.includes(config.set)) {
+      await this.createEnvironment(config);
+    }
+    let featureFlag;
+    try {
+      featureFlag = await this.getFeatureFlag(config);
+      const existingVariation =
+        featureFlag.variations.find((variation: any) => variation.value === config.value) ||
+        (await this.createVariation(config));
+      if (existingVariation) featureFlag = await this.updateDefaultFallthroughVariation(config, existingVariation._id);
+    } catch (e) {
+      if (e.response.status === 404) featureFlag = await this.createFeatureFlag(config);
+      else throw e;
+    }
+    if (featureFlag && !featureFlag.environments[config.set].on) {
+      const onData = {
+        environmentKey: config.set,
+        instructions: [{ kind: 'turnFlagOn' }],
+      };
+      await this.patchUpdate(config, onData);
+    }
+  }
+
+  private async validateQueries(queries: ConfigStoreQuery[]): Promise<[string, string[]]> {
+    await this.updateEnvironments();
+    queries.forEach((query) => {
+      assert(query.set.split('/').length === 1, 'Subsets are not supported in LaunchDarklyStore');
+    });
+    const noRootQueries = queries.filter((query) => query.set !== '' && this.environments.includes(query.set));
+    if (noRootQueries[0]) {
+      const env = noRootQueries[0].set;
+      const keys = noRootQueries.map((value) => value.key);
+      return [env, keys];
+    }
+    throw new Error('No valid queries');
+  }
+
+  private assertNotSettingToRoot(configs: Config[]) {
+    assert.strictEqual(
+      configs.some((config) => config.set === ''),
+      false,
+      'ConfigSet cannot be empty string',
+    );
+  }
```

Comment:
These methods should be merged into a single method that validates sets. Any invalid set should cause the operation to throw. The validation method shouldn't return anything, it should only validate.

Is there a reason why in `validateQueries` you don't validate and throw if one of the sets is the root set?

Reply:
I'm "pretty sure" line 130 filters root-set from the queries.

They way EvalCommand composing these queries is hierarchical. So it's certain that this method will receive query for root-set.
If it's not clear i'll show you in-person in the code.
This is also the reason these are separated.

I will change the name for `validateQueries` to whatever you'd say but it will always return the set and the keys.

BTW, the only reason to use `if` and `throw` instead of `assert` here is TS giving me errors that don't exist.